lib/moab/storage_object_validator.rb

```diff
@@ -98,6 +102,13 @@ def check_correctly_formed_moabs
           errors.concat check_sub_dirs(data_sub_dirs, version_dir, EXPECTED_DATA_SUB_DIRS)
           errors.concat check_required_manifest_files(version_path, version_dir)
         end
+        if before_result_size == errors.size
```

Comment:
can these statements be added to the if block above?  if not, then isn't the if condition wrong?

Reply:
Well i guess here, In the first if block 
``` 
if before_result_size == after_result_size
          data_dir_path = "#{version_path}/#{DATA_DIR_NAME}"
          data_sub_dirs = sub_dirs(data_dir_path)
          errors.concat check_sub_dirs(data_sub_dirs, version_dir, EXPECTED_DATA_SUB_DIRS)
          errors.concat check_required_manifest_files(version_path, version_dir)
        end
```

I want this to run if the sizes are equal, meaning nothing has been added to the errors array. Now if this if loop is executed then the errors array will either stay the same size or get bigger. Now if it stays the same size i then want the next if loop to run ( ``` if before_result_size == errors.size
          x = []
          metadata_dir_path = "#{version_path}/#{DATA_DIR_NAME}/#{METADATA_DIR_NAME}"
          metadata_sub_dir = sub_dirs(metadata_dir_path)
          metadata_sub_dir.each { |item| x << File.directory?("#{metadata_dir_path}/#{item}") }
          errors << result_hash(DIRS_DETECTED) if x.include?(true)
        end ``` ) but if the errors array is bigger then i don't want the 2nd if loop to run

Reply:
I'm sure there is a better way of coding this then 2 if loops.

Reply:
i'm trying to think of a better way to do this

Reply:
Just noting that @SaravShah and I spoke F2F and are thinking an approach like this might be clearer:

```
if before_result_size == after_result_size
      data_dir_path = "#{version_path}/#{DATA_DIR_NAME}"
      data_sub_dirs = sub_dirs(data_dir_path)
      errors.concat check_sub_dirs(data_sub_dirs, version_dir, EXPECTED_DATA_SUB_DIRS)
      errors.concat check_required_manifest_files(version_path, version_dir)
      if before_result_size == errors.size
          x = []
          metadata_dir_path = "#{version_path}/#{DATA_DIR_NAME}/#{METADATA_DIR_NAME}"
          metadata_sub_dir = sub_dirs(metadata_dir_path)
          metadata_sub_dir.each { |item| x << File.directory?("#{metadata_dir_path}/#{item}") }
          errors << result_hash(DIRS_DETECTED) if x.include?(true)
        end  
end
```  

It might make sense to make the inner conditional code into a well-named method ...

Note that the current implementation is NOT recomputing after_result_size so won't it always do both conditionals, which is incorrect (there should be a test for this in both directions, no?).